File: postgremq-ts/src/__tests__/integration/schema.test.ts
import { test, expect } from '@jest/globals';
import { Connection } from '../../connection';
import { createIsolatedTestConnection, getSharedTestDatabase } from '../helpers';

test('application transactions publish and ack atomically with a separate queue schema', async () => {
  const iso = await createIsolatedTestConnection(await getSharedTestDatabase());
  const c = iso.connection;
  let consumer: ReturnType<Connection['consume']> | undefined;
  try {
    await iso.pool.query('CREATE SCHEMA app');
    for (const table of ['topics', 'queues', 'messages', 'queue_messages', 'dead_letter_queue']) {
      await iso.pool.query(`CREATE TABLE app.${table} (value text)`);
    }
    await iso.pool.query(`
      CREATE FUNCTION app.publish_message(varchar, jsonb) RETURNS bigint
      LANGUAGE plpgsql AS $$ BEGIN RAISE EXCEPTION 'application publish called'; END $$;
      CREATE FUNCTION app.ack_message(varchar, bigint, varchar) RETURNS void
      LANGUAGE plpgsql AS $$ BEGIN RAISE EXCEPTION 'application ack called'; END $$;
    `);
    await c.createTopic('in');
    await c.createTopic('out');
    await c.createQueue('q', 'in', false);
    const incoming = await c.publish('in', {});
    consumer = c.consume('q', {
      batchSize: 1,
      pollingIntervalMs: 20,
      visibilityTimeoutSec: 30,
      autoExtension: { enabled: false },
    });
    const messages = consumer.messages();
    for (const commit of [false, true]) {
      const msg = (await messages.next()).value;
      expect(msg.id).toBe(incoming);
      const tx = await iso.pool.connect();
      try {
        await tx.query('BEGIN');
        await tx.query('SET LOCAL search_path = app, pg_temp');
        await tx.query("INSERT INTO messages(value) VALUES ('application data')");
        const outgoing = await c.publishWithTransaction(tx, 'out', {});
        await msg.ackWithTransaction(tx);
        expect((await tx.query('SHOW search_path')).rows[0].search_path).toBe('app, pg_temp');
        const check = async (count: number, state: string) => {
          expect((await iso.pool.query('SELECT * FROM app.messages')).rows).toHaveLength(count);
          expect(
            (await iso.pool.query('SELECT * FROM postgremq.messages WHERE id=$1', [outgoing])).rows
          ).toHaveLength(count);
          expect(
            (
              await iso.pool.query(
                'SELECT status FROM postgremq.queue_messages WHERE message_id=$1',
                [incoming]
              )
            ).rows[0].status
          ).toBe(state);
        };
        await check(0, 'processing');
        await tx.query(commit ? 'COMMIT' : 'ROLLBACK');
        await check(commit ? 1 : 0, commit ? 'completed' : 'processing');
        if (!commit) {
          await iso.pool.query(
            "UPDATE postgremq.queue_messages SET vt=clock_timestamp()-interval '1 second' WHERE message_id=$1",
            [incoming]
          );
        }
      } finally {
        await tx.query('ROLLBACK');
        tx.release();
      }
    }
    await consumer.stop();
  } finally {
    await c.close();
    await iso.dropDatabase();
  }
});

File: postgremq-go/schema_test.go
package postgremq_go_test

import (
	"fmt"
	"testing"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	postgremq "github.com/slavakl/postgremq/postgremq-go"
	"github.com/stretchr/testify/require"
)

func TestSchemaIsolationWithApplicationTransactions(t *testing.T) {
	base, ctx := setupEmptyTestDatabase(t)
	_, err := base.Exec(ctx, "CREATE SCHEMA app")
	require.NoError(t, err)
	for _, name := range []string{"topics", "queues", "messages", "queue_messages", "dead_letter_queue"} {
		_, err = base.Exec(ctx, fmt.Sprintf("CREATE TABLE app.%s (value text)", name))
		require.NoError(t, err)
	}
	_, err = base.Exec(ctx, `
		CREATE TABLE app.postgremq_migrations (version bigint, dirty boolean);
		INSERT INTO app.postgremq_migrations VALUES (99, true);
		CREATE FUNCTION app.publish_message(varchar, jsonb) RETURNS bigint
		LANGUAGE plpgsql AS $$ BEGIN RAISE EXCEPTION 'application publish called'; END $$;
		CREATE FUNCTION app.ack_message(varchar, bigint, varchar) RETURNS void
		LANGUAGE plpgsql AS $$ BEGIN RAISE EXCEPTION 'application ack called'; END $$;
	`)
	require.NoError(t, err)
	cfg := base.Config()
	cfg.ConnConfig.RuntimeParams["search_path"] = "app, pg_temp"
	pool, err := pgxpool.NewWithConfig(ctx, cfg)
	require.NoError(t, err)
	defer pool.Close()

	status, err := postgremq.GetMigrationStatus(pool)
	require.NoError(t, err)
	require.Zero(t, status.CurrentVersion)
	var exists bool
	require.NoError(t, pool.QueryRow(ctx, "SELECT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname='postgremq')").Scan(&exists))
	require.False(t, exists, "status must not create the queue schema")
	require.NoError(t, postgremq.Migrate(pool, postgremq.MigrateOptions{}))
	status, err = postgremq.GetMigrationStatus(pool)
	require.NoError(t, err)
	require.Equal(t, uint(1), status.CurrentVersion)
	require.False(t, status.Dirty)
	var version int
	require.NoError(t, pool.QueryRow(ctx, "SELECT version FROM app.postgremq_migrations").Scan(&version))
	require.Equal(t, 99, version)
	var publicObjects int
	require.NoError(t, pool.QueryRow(ctx, `SELECT
		(SELECT count(*) FROM pg_class WHERE relnamespace='public'::regnamespace) +
		(SELECT count(*) FROM pg_proc WHERE pronamespace='public'::regnamespace)`).Scan(&publicObjects))
	require.Zero(t, publicObjects)

	c, err := postgremq.DialFromPool(pool, postgremq.WithShutdownTimeout(time.Second))
	require.NoError(t, err)
	defer c.Close()
	require.NoError(t, c.CreateTopic(ctx, "in"))
	require.NoError(t, c.CreateTopic(ctx, "out"))
	require.NoError(t, c.CreateQueue(ctx, "q", "in", false))
	incoming, err := c.Publish(ctx, "in", []byte(`{}`))
	require.NoError(t, err)
	consumer, err := c.Consume("q", postgremq.WithBatchSize(1), postgremq.WithVT(30), postgremq.WithNoAutoExtension(), postgremq.WithCheckTimeout(20*time.Millisecond))
	require.NoError(t, err)

	for _, commit := range []bool{false, true} {
		var msg *postgremq.Message
		select {
		case msg = <-consumer.Messages():
		case <-time.After(3 * time.Second):
			t.Fatal("message not delivered")
		}
		require.Equal(t, incoming, msg.ID)
		tx, err := pool.Begin(ctx)
		require.NoError(t, err)
		defer tx.Rollback(ctx)
		_, err = tx.Exec(ctx, "INSERT INTO messages(value) VALUES ('application data')")
		require.NoError(t, err)
		outgoing, err := c.PublishWithTx(ctx, tx, "out", []byte(`{}`))
		require.NoError(t, err)
		require.NoError(t, msg.AckWithTx(ctx, tx))
		var path string
		require.NoError(t, tx.QueryRow(ctx, "SHOW search_path").Scan(&path))
		require.Equal(t, "app, pg_temp", path)
		var appRows, published int
		var state string
		check := func(wantRows int, wantState string) {
			t.Helper()
			require.NoError(t, pool.QueryRow(ctx, "SELECT count(*) FROM app.messages").Scan(&appRows))
			require.NoError(t, pool.QueryRow(ctx, "SELECT count(*) FROM postgremq.messages WHERE id=$1", outgoing).Scan(&published))
			require.NoError(t, pool.QueryRow(ctx, "SELECT status FROM postgremq.queue_messages WHERE message_id=$1", incoming).Scan(&state))
			require.Equal(t, wantRows, appRows)
			require.Equal(t, wantRows, published)
			require.Equal(t, wantState, state)
		}
		check(0, "processing")
		if commit {
			require.NoError(t, tx.Commit(ctx))
			check(1, "completed")
		} else {
			require.NoError(t, tx.Rollback(ctx))
			check(0, "processing")
			_, err = pool.Exec(ctx, "UPDATE postgremq.queue_messages SET vt=clock_timestamp()-interval '1 second' WHERE message_id=$1", incoming)
			require.NoError(t, err)
		}
	}
}
